FEA Add `EngineAwareMixin` to factor the common logic of estimators which are plugin-extendable

betatim opened the pull request:
Adds a mixin for estimators to use when they support having an engine. 

Also adds a decorator that can be used on `fit` to convert the estimators attributes to either `engine_native` or `sklearn_native` types.

WDYT?

Notes: I considered using something like `__init_subclass__` to automagically wrap `fit` but decided that "explicit is better than implicit", so this is how we got the decorator approach. After building something based on `__init_subclass__`, which seemed cool because it means all you have to do is add the mixin, it seemed a bit too magical that this would wrap `fit`. So I switched to having a decorator.

fcharras commented:
Sorry I wasn't aware of this PR until now. Will take a look !

ogrisel commented on sklearn/_engine/base.py:
```diff
@@ -112,3 +112,29 @@ def get_engine_classes(engine_name, default, verbose=False):
+        if convert_attributes == "sklearn_types":
+            engine = self._engine_class
+            for name, attribute in vars(self).items():
+                # Using __dlpack__ as signal for "not a basic Python type"
+                if hasattr(attribute, "__dlpack__"):
```
```suggestion
                # Using __dlpack__ as signal for "not a basic Python type"
                if hasattr(attribute, "__dlpack__"):
```

Maybe this choice (to convert or passthrough) should be left to the engine itself because it can know the specifics of the estimator better than the generic decorator (in particular by using the name).

betatim replied:
I like that suggestion. I ended up using `__dlpack__` because I couldn't find anything good. But I didn't think of just passing everything to the engine and leaving it up to it to decide.

ogrisel commented on sklearn/_config.py:
```diff
@@ -260,6 +274,15 @@ def config_context(
+    engine_attributes : str, default=None
+        Enable conversion of estimator attributes to scikit-learn native
+        types by setting to "sklearn_types". By default attributes are
+        stored using engine native types.
```
```suggestion
        stored using engine native types so as to avoid wasteful
        data container conversions and memory transfers between host
        and device between the `fit` and `predict`/`transform` calls
        of an engine-aware estimator.
```

ogrisel commented on sklearn/_engine/tests/test_engines.py:
```diff
@@ -136,3 +140,32 @@ def test_get_engine_class():
+@pytest.mark.parametrize(
+    "attribute_types,converted", [("sklearn_types", True), ("engine_types", False)]
+)
+def test_attribute_conversion(attribute_types, converted):
```
I think some older versions of numpy still supported by scikit-learn my not expose the `np.array_api` module and we should skip this test in that case.

ogrisel commented on sklearn/_engine/tests/test_engines.py:
```diff
@@ -136,3 +140,32 @@ def test_get_engine_class():
+    class Estimator:
+        _engine_class = Engine
+
+        @convert_attributes
+        def fit(self, X):
+            self.x = np.array_api.asarray(X)
+
+    X = np.array([1, 2, 3])
+    est = Estimator()
+    with config_context(engine_attributes=attribute_types):
+        est.fit(X)
+
+    assert isinstance(est.x, np.ndarray) == converted
```
For the sake of naming conventions:

```suggestion
        @convert_attributes
        def fit(self, X):
            self.X_ = np.array_api.asarray(X)

    X = np.array([1, 2, 3])
    est = Estimator()
    with config_context(engine_attributes=attribute_types):
        est.fit(X)

    assert isinstance(est.X_, np.ndarray) == converted
```

ogrisel reviewed (approved):
This looks great! Just a few minor suggestions below:

betatim commented:
Implemented the suggestions almost exactly as suggested. Made a change to the docstring suggestion and renamed the conversion function (not sure I like the name but it seems better than "convert to numpy", when really it is "to sklearn types").

betatim commented:
@ogrisel what do you think about the changes in the last commit? This makes it possible to perform input validation in the "acceptance" function. Ideas on making the return value better (I dislike the `True, input_vals_dict` thing. Maybe the engine should store the values on its instance (and we don't pass them in to `prepare_fit` and so on?)? That would make it clear that the engine can do what it wants with the values (modify them, convert them, etc) and that they won't change between calling `validate` and the other steps of fitting.

ogrisel commented on sklearn/_engine/base.py:
```diff
@@ -112,3 +112,37 @@ def get_engine_classes(engine_name, default, verbose=False):
+
+        if convert_attributes == "sklearn_types":
+            engine = self._engine_class
+            for name, attribute in vars(self).items():
+                # All attributes are passed to the engine, which can
+                # either convert the value (engine specific types) or
+                # return it as is (native Python types)
+                converted = engine.convert_to_sklearn_types(name, attribute)
```
```suggestion
            for name, value in vars(self).items():
                # All attributes are passed to the engine, which can
                # either convert the value (engine specific types) or
                # return it as is (native Python types)
                converted = engine.convert_to_sklearn_types(name, value)
```

ogrisel replied:
We could additionally allow the engine to implement an optional `engine.convert_estimator_to_sklearn_types` method if we want to give it the flexibility to do the attribute looping and `setattr` logic by itself.

But this probably qualifies as a YAGNI for now. Let's just keep this possibility in mind for later if find the default order returned by `vars(self).items()` annoying for some estimators.

ogrisel replied:
Another potential limitation of this design would be the handling of the case when the `EngineAwareMixin` class is used on a meta-estimator. But let's keep this for later.

betatim replied:
> Another potential limitation of this design would be the handling of the case when the `EngineAwareMixin` class is used on a meta-estimator. But let's keep this for later.

Just to double check: you are thinking of the estimator(s) that are stored as attributes of the meta-estimator right?

ogrisel commented on sklearn/cluster/_kmeans.py:
```diff
@@ -1572,14 +1593,16 @@ def fit(self, X, y=None, sample_weight=None):
-        engine = self._get_engine(X, y, sample_weight, reset=True)
+        engine, validated_inputs = self._get_engine(X, y, sample_weight, reset=True)
 
+        X = validated_inputs["X"]
+        y = validated_inputs["y"]
+        sample_weight = validated_inputs["sample_weight"]
```
Too magic?

```suggestion
        # Re-assign the X, y and sample_weight local variables
        locals().update(validated_inputs)
```

no opinion, feel free to mark as resolved if you don't like it but this would avoid quite some verbose duplication.

jjerphan replied:
What if `validated_inputs` contains more than those 3 items? Do we want to store references to those in/via `locals()`?

ogrisel replied:
I think it's the responsibility of the plugin to not mess around with the estimator's internal.

ogrisel commented on sklearn/cluster/_kmeans.py:
```diff
@@ -275,12 +275,49 @@ class KMeansCythonEngine:
+        Should fail as quickly as possible.
+        """
+        # The default engine accepts everything and does not convert inputs
+        return True, {"X": X, "y": y, "sample_weight": sample_weight}
```
I have the feeling that we should factor in the calls to `self.estimator._validate_data` and remove the `_check_test_data` method and simplify the `_prepare_fit` method.

Otherwise, the default engine would not behave in a consistent way with non-default engine and therefore would not serve as an educational implementation template for third party engine implementers.

ogrisel commented on sklearn/base.py:
```diff
@@ -990,6 +992,38 @@ def _more_tags(self):
+                "X": X,
+                "y": y,
+                "sample_weight": sample_weight,
+            }
```
If the above suggestion is implemented, we should call `engine.validate` here.

betatim replied:
I thought about that but then left it out. My thinking was that "once the engine has been selected during fit it will be selected forever (until re-fit)". `validate` is used as part of the selection, so why call it again?

However, a reason to call it again is that it performs input conversion. But then we should call it before `fit` and `predict`. Which then makes me wonder what the purpose of `prepare_fit` is (right now it does lots of conversion).

For example in the `KMeans` case `_validate_data` is called from `prepare_fit` and from `prepare_prediction`, but with different arguments (`reset=False` vs `reset=True`). So if we move it to `accept` we need to pass information to `accept` about which context it is being called in (fit or predict). So maybe there should be `accept_fit` and `accept_predict`? But then what is the difference to `prepare_fit` and `prepare_predict`?

Overall it somehow feels like the API isn't "nice" :-/

fcharras replied:
The engine definitely needs to validate and convert the input that the user submits predict methods in some way, IMO it's a core requirement to keep this behavior. Regarding how, I agree that it doesn't speak for itself. It seems that the existing implementation took the most straightforward path to factor identical steps in common engine methods but let's change it if there's another better possible abstraction.

ogrisel replied:
> For example in the KMeans case _validate_data is called from prepare_fit and from prepare_prediction, but with different arguments (reset=False vs reset=True). So if we move it to accept we need to pass information to accept about which context it is being called in (fit or predict). So maybe there should be accept_fit and accept_predict?

Maybe we can try to just pass the `reset` argument to the `accept` method for a start? Or alternatively pass a `context="fit"/"predict"/"transform"` string.

Then we can probably remove the `prepare_` methods.

betatim replied:
After our discussion on Wednesday the PR is now back to using an `accept()` that can't modify the input data. That job is left for the `prepare_*` methods.

ogrisel reviewed (commented):
Thanks for the update @betatim. Here is a first pass of feedback.

I would also love to hear from @fcharras on the API design proposed by team and my own suggestions.

ogrisel commented on sklearn/_engine/base.py:
```diff
@@ -112,3 +112,37 @@ def get_engine_classes(engine_name, default, verbose=False):
+            # conversion to the sklearn native types the default engine
+            # is used.
+            self._engine_class = self._default_engine
+            self._engine_provider = "default"
```
We should probably test this in the test.

jjerphan commented on sklearn/cluster/_kmeans.py:
```diff
@@ -275,12 +275,49 @@ class KMeansCythonEngine:
+        converted
+            Attribute value converted to a scikit-learn native type.
+        """
+        # XXX Maybe a bit useless as it should never get called, but it
+        # does demonstrate the API
+        return value
```
LGTM for now. I think this is a no-op generally.

jjerphan commented on sklearn/_engine/base.py:
```diff
@@ -112,3 +112,37 @@ def get_engine_classes(engine_name, default, verbose=False):
         yield provider, engine_class
+
+
+def convert_attributes(method):
```
Note for future iteration: Could method decorated by `convert_attributes` be called concurrently? If it is the case, should we add some handling to make converting attributes safe?

jjerphan commented on sklearn/cluster/_kmeans.py:
```diff
@@ -275,12 +275,49 @@ class KMeansCythonEngine:
+
+        Should fail as quickly as possible.
+        """
+        # The default engine accepts everything and does not convert inputs
+        return True, {"X": X, "y": y, "sample_weight": sample_weight}
```
```suggestion
        # The default engine accepts everything and does not convert inputs
        accept_inputs = True
        return accept_inputs, {"X": X, "y": y, "sample_weight": sample_weight}
```

ogrisel replied:
It's not just about accepting the inputs but the inputs in conjunction with the estimator hyper-parameters.

I think it's ok for an engine to accept to be enabled only for some hyper-parameters, even if the inputs are all valid (e.g. depending of the `estimator.algorithm` parameter for `KMeans`, lloyd vs elkan).

So I would rather keep the short `accept` variable name.

jjerphan replied:
Then:
```suggestion
        # The default engine accepts everything and does not convert inputs
        accept = True
        return accept, {"X": X, "y": y, "sample_weight": sample_weight}
```

jjerphan reviewed (approved):
Thanks, @betatim.

At this stage, this LGTM. I agree with @ogrisel's remarks and have a few comments.

fcharras commented:
Thanks for the work @betatim 

I also caught up with the PR, I like the overall approach, don't have anything to add beside what's has already been suggested. 

In particular like @ogrisel suggested I think `_get_engine` is supposed to returned input validated by the engine even when `hasattr(self, "_engine_provider") and not reset:` ? or is there another reason behind not validating ? (it seems like an error since later on the prediction methods expect validated output already)

When it's merged I'll open a PR on our gpu plugin to see how it goes !

ogrisel commented:
Just to make sure that this is visible, I replied in the previous thread:

https://github.com/ogrisel/scikit-learn/pull/13#discussion_r1073379037

betatim commented:
I've updated the code so that `accept` does not return the inputs. We discussed this two days ago and settled on having a separate "acceptance" and "conversion/validation" phase.

To test it I also added a first draft of being able to pass an engine class to `config_context(engine_provider=[SomeEngineAsAClass, "some_other_plugin_name"])`. This way you can define an engine at runtime and pass it in. This solves the problem that you can't define a new engine at runtime, because you can't register a new entry point.

ogrisel commented on sklearn/cluster/_kmeans.py:
```diff
@@ -275,15 +275,51 @@ class KMeansCythonEngine:
+        """Validate input and hyper-parameters.
+
+        Determine if the engine can handle the hyper-parameter of the estimator
+        as well as the input data.
```
```suggestion
        as well as the input data. If not, this would allow to select another
        engine positioned after in the current provider list or ultimately
        the default engine of the estimator as fallback.
```

betatim replied:
I edited the docstring to incorporate what you suggested but in a slightly different way.

I feel like the explanation of how the engine selection works somehow should be somewhere else no? Somewhere that "defines" the engine interface. Maybe something to explain in more detail in the documentation (I don't think we should introduce some kind of base engine class just to define the API, feels wrong.)

ogrisel commented on sklearn/_engine/base.py:
```diff
@@ -91,16 +97,22 @@ def _get_engine_classes(engine_name, provider_names, engine_specs, default):
-        provider_names = tuple(provider_names)
+    # Single provider class was passed in
+    elif inspect.isclass(provider_names):
+        provider_names = (provider_names,)
```
```suggestion
        # allow direct registration of engine classes to ease testing, debugging
        # and benchmarking without having to register a fake package with metadata
        # just to use a custom engine not meant to be used by end-users.
        provider_names = (provider_names,)
```

betatim replied:
Implemented this, but I moved the code to `_config.py`. Moved it because I didn't want to have to duplicate this logic in `base.py` and `_engine/base.py` which both access this config value.

What is the general thinking on having such "processing" logic in `_config.py`?

ogrisel commented on sklearn/_config.py:
```diff
@@ -133,6 +135,17 @@ def set_config(
 
```
Could you please update the `engine_provider` entry above to mention that it's possible to either register the name of an entry point or directly pass an engine class instance for testing purpose?

Or do we want to keep the engine class backdoor "secret" for now?

ogrisel replied:
Let's do this for now.

Since this API is explicitly experimental we can still change our mind later.

ogrisel reviewed (commented):
Thanks, this looks good to me. I would just like to better define the case when `reset=False` with a previously resolved engine class as explained below.

ogrisel commented on sklearn/_engine/tests/test_engines.py:
```diff
@@ -136,3 +140,82 @@ def test_get_engine_class():
+            engine = self._get_engine(X, y, reset=True)
+            return engine
+
+    with config_context(engine_provider=(NeverAcceptsEngine)):
```
Misleading parens:

```suggestion
    with config_context(engine_provider=NeverAcceptsEngine):
```

ogrisel commented on sklearn/_engine/tests/test_engines.py:
```diff
@@ -136,3 +140,82 @@ def test_get_engine_class():
+        engine = est.resolve_engine([[1, 2], [3, 4]], [0, 1])
+        assert isinstance(engine, DefaultEngine)
+
+    with config_context(engine_provider=(AlwaysAcceptsEngine)):
```
```suggestion
    with config_context(engine_provider=AlwaysAcceptsEngine):
```

fcharras commented:
So we've completed a circle :grin: 

It would help if you can highlight how it's intended to be used with a few examples regarding the inputs that could reasonably be inputed by a user and limit use-cases in the range of supported inputs (list of lists, input dtype that might require casting depending on the engine,...). What would you recommend to plugin developers regarding the behavior of the engine in each case (~does `accept` return `True` or `False`, does `prepare_*` raise an error and which kind of error).

If it's back to the same `accept` as a plugin developer I'm still tempted to come back to https://github.com/soda-inria/sklearn-numba-dpex/pull/74 (where `accept` uses `asarray` and store the converted array before returning `True`) but improving it so that it raises errors for cases where we can be sure other engines will all fail too (e.g `ndim`).

If you consider it's a bad practice, I'd suggest to throw the `engine` instance after `accept` and re-instantiate later. But then, `acceptance` criterion should be expected to be light, because I'd see triggering a conversion just for acceptation a bit wasteful.

ogrisel commented:
> If it's back to the same accept as a plugin developer I'm still tempted to come back to https://github.com/soda-inria/sklearn-numba-dpex/pull/74 (where accept uses asarray and store the converted array before returning True) but improving it so that it raises errors for cases where we can be sure other engines will all fail too (e.g ndim).

I would rather not do store the attribute. I think in the case of `sklearn_numba_dpex`, it should only return `False` if `scipy.sparse.issparse(X)` and return `True` otherwise. Then then the validation itself would happen in the `prepare_` methods as usual and raise exceptions in case of problems.

Later we might refine the `accept` logic of `sklearn_numba_dpex` engines so that, for inputs which have a `__dlpack__` method we inspect the device and it it's not a device managed by the sycl runtime, they would return `False`.

fcharras commented:
> I think in the case of `sklearn_numba_dpex`, it should only return `False` if the `algorithm` param is not Lloyd or if `scipy.sparse.issparse(X)` and return `True` otherwise. Then then the validation itself would happen in the `prepare_` methods as usual and raise exceptions in case of problems.

:+1: I'm fine with that. :+1:

jjerphan commented:
Small interjection: Do you think the following would be a better title for this PR ?

```
FEA Add `EngineAwareMixin` to factor the common logic of estimators with are plugin-extendable
```

betatim commented on sklearn/_engine/tests/test_engines.py:
```diff
@@ -136,3 +140,82 @@ def test_get_engine_class():
+
+    with config_context(engine_provider=(NeverAcceptsEngine)):
+        est = Estimator()
+        engine = est.resolve_engine([[1, 2], [3, 4]], [0, 1])
```
Good idea, I don't know why I didn't think of that :-/

Also added tests for `reset=True` and `False`.

betatim commented on sklearn/base.py:
```diff
@@ -990,6 +992,31 @@ def _more_tags(self):
+        again.
+        """
+        if hasattr(self, "_engine_provider") and not reset:
+            return self._engine_class(self)
```
Good point. Done.

betatim commented:
> Small interjection: Do you think the following would be a better title for this PR ?
> 
> ```
> FEA Add `EngineAwareMixin` to factor the common logic of estimators which are plugin-extendable
> ```

That is indeed a better title. Switched to it. However, I think this PR is becoming a collection of several ideas (e.g. it "accidentally" also implements the idea of having ad-hoc engines that aren't registered via entrypoints), so IMHO we should try and wrap this one up and start new PRs for new ideas, instead of putting them all here.

ogrisel commented on sklearn/_engine/base.py:
```diff
@@ -82,6 +84,14 @@ def _get_engine_classes(engine_name, provider_names, engine_specs, default):
+            # we use the class itself. This mirrors what the user used
+            # when they set the config (ad-hoc class or string naming
+            # a provider).
+            yield provider_name, provider_name
```
```suggestion
            engine_class = provider_name
            yield f"{engine_class.__module__}.{engine_class.__name__}", engine_class
```

ogrisel commented on sklearn/_engine/tests/test_engines.py:
```diff
@@ -23,6 +27,40 @@ class NestedFakeEngine:
 
 
+# Dummy classes used to test engine resolution
+class DefaultEngine:
```
```suggestion
class DefaultEngine:
    engine_name = "test-engine"
```

ogrisel commented on sklearn/_engine/tests/test_engines.py:
```diff
@@ -23,6 +27,40 @@ class NestedFakeEngine:
+        return True
+
+
+class NeverAcceptsEngine:
```
```suggestion
class NeverAcceptsEngine:
    "test-engine"
```

ogrisel commented on sklearn/_engine/tests/test_engines.py:
```diff
@@ -23,6 +27,40 @@ class NestedFakeEngine:
+        return False
+
+
+class AlwaysAcceptsEngine:
```
```suggestion
class AlwaysAcceptsEngine:
    engine_name = "test-engine"
```

ogrisel reviewed (approved):
LGTM with the previous and following suggestion for class filtering by engine_name.
